### ISSUE TITLE
Questions: Compatibility with Hive, Spark SQL?

### ISSUE THREAD

**Issue by thomaskwan** [alanmcruickshank/sqlfluff#417]
Wonder if anyone is using this project with Hive and Spark SQL? Can users implement their own dialects? I do not see references on how to inject our own rules. Is this possible?

thanks for your time. very good project

**Comment by alanmcruickshank** [comment]
Hi @thomaskwan - yes we can totally implement new dialects. I don't know if anyone has tried with Spark or Hive, but there's no structural reason that it shouldn't work. Are you able to post a few example queries here to check? One or two simple ones and then one or two complex ones would be great to assess how much work it would be to get full compatability.

As an example, @Katzmann1983 implemented most of the teradata dialect as one person, so it's very doable to add a new dialect.

**Comment by satish-ravi** [comment]
Hi, I tried running this with Hive, here are few example queries that failed with parsing:

```sql
CREATE TABLE foo (field1 string, field2 string) PARTITIONED BY (field3 string);
INSERT OVERWRITE TABLE foo SELECT field1, field2, field3 FROM bar WHERE field4 = 'foo';
SELECT * FROM ${hiveconf:table_name};
```

Here are a couple of complex queries:
https://gist.github.com/satish-ravi/d2b4b7d6401961f03fd8d7087c74cfd7
https://gist.github.com/satish-ravi/4a169ed27c1ef6127f2e85133848b727

Here's the full language specification: https://cwiki.apache.org/confluence/display/Hive/LanguageManual

I am happy to contribute a dialect for Hive with some guidance.

Thanks. Very cool project.

**Comment by dmateusp** [comment]
hi @satish-ravi you can join us on Slack https://join.slack.com/t/sqlfluff/shared_invite/zt-js0wsnvo-78gEI0phW9_LiItXAiJcqQ for more synchronous communication if you'd wish, to get you started implementing a new Dialect :)

Here's some pointers:
* Platform specific dialects are all derived from a base dialect: the ANSI dialect: https://github.com/sqlfluff/sqlfluff/blob/master/src/sqlfluff/core/dialects/dialect_ansi.py.
* You can read https://github.com/sqlfluff/sqlfluff/blob/master/src/sqlfluff/core/dialects/dialect_snowflake.py to have an example of a dialect implementation (here for Snowflake).
* SQLFluff recursively parses SQL into chunks which are called Segments in the code. You can see that in action by using the very verbose parsing option `sqlfluff parse -vvv some_sql_file.sql`. There's an explanation of how the templater, lexer and parser work here: https://docs.sqlfluff.com/en/latest/architecture.html.
* Parser/dialect tests ran through https://github.com/sqlfluff/sqlfluff/blob/master/test/core/dialects/dialects_test.py, and the fixtures are added here: https://github.com/sqlfluff/sqlfluff/tree/master/test/fixtures/parser.

As an aside, I wonder if HiveQL and Spark SQL are similar enough that one can be derived from the other (to limit redundant code) 🤔 , but we should have one of them working first :)

**Comment by satish-ravi** [comment]
Thanks @dmateusp, I'll check these out and get started. Re HiveQL and spark SQL, I think it should be possible to derive one from the other. I plan to get started on HiveQL.

**Comment by foundinblank** [comment]
Hi - we use SparkSQL so we'd definitely be interested in any work you've done so far with a dialect for Hive/Spark SQL. Found this issue via this Slack thread: https://sqlfluff.slack.com/archives/C01GPHR8J0G/p1618571344021200

**Comment by satish-ravi** [comment]
I was finally able to spend some time on this and added a hive dialect with DDL statements (PR #985). Will continue adding other statements in the next couple of weeks.

**Comment by lpillmann** [comment]
Hey all! Just wanted to say I'm glad someone has worked on this. Kudos to @satish-ravi :)

Really looking forward to start using sqlfluff in our huge HQL codebase!

**Comment by R7L208** [comment]
I'm setting up CI for a spark3 SQL project and started working on a spark3 dialect a few days ago. SQLFluff is great, and I am very excited about the interest in HQL and Spark dialects :)

Spark generally maintains a native and Hive based format for commands. Since the Spark project is moving towards ANSI compliance overall, I think the best option would be to inherit from the ANSI dialect then use `get_segments` and `get_grammar` to retrieve the needed Hive format commands and common syntax that isn't shared with ANSI (i.e. partitioning). In my opinion, basing spark off of the hive dialect would cause more work in the future as these continue to diverge.
